cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      if (scheduleTask == null) {
+        throw new ScheduleNotFoundException(scheduleName);
+      }
+      if (scheduleTask.isActive()) {
+        activeTasks.decrementAndGet();
+      }
+    }
+
+    /**
+     * Get the status a scheduling task that is based on the data received by the stream referenced by {@code this}
+     * object.
+     */
+    public ScheduleState scheduleTaskState(Id.Program programId, SchedulableProgramType programType,
+                                           String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null) {
+        return ScheduleState.NOT_FOUND;
+      }
+      return task.isActive() ? ScheduleState.SCHEDULED : ScheduleState.SUSPENDED;
+    }
+
+    /**
+     * @return true if this object does not reference any schedule, false otherwise
+     */
+    public boolean isEmpty() {
+      return scheduleTasks.isEmpty();
+    }
+
+    public Id.Stream getStreamId() {
+      return streamId;
+    }
+
+    @Override
+    public Type getNotificationFeedType() {
+      return StreamSizeNotification.class;
+    }
+
+    @Override
+    public void received(StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only pass the stream size notification to the schedule tasks if the notification
+      // came after the last seen notification
+      boolean send = false;
+      if (activeTasks.get() > 0) {
+        cancelPollingAndScheduleNext();
+      }
+      synchronized (lastNotificationLock) {
+        if (lastNotification == null || notification == lastNotification ||
+          notification.getTimestamp() > lastNotification.getTimestamp()) {
+          send = true;
+          lastNotification = notification;
+        }
+      }
+      if (send) {
+        sendNotificationToActiveTasks(notification);
+      }
+    }
+
+    /**
+     * Send a {@link StreamSizeNotification} to all the active {@link StreamSizeSchedule} referenced
+     * by this object.
+     */
+    private void sendNotificationToActiveTasks(final StreamSizeNotification notification) {
+      for (final StreamSizeScheduleTask task : scheduleTasks.values()) {
+        if (!task.isActive()) {
+          continue;
+        }
+        notificationExecutor.execute(new Runnable() {
+          @Override
+          public void run() {
+            task.received(notification);
+          }
+        });
+      }
+    }
+
+    private Id.NotificationFeed getFeed() {
+      return new Id.NotificationFeed.Builder()
+        .setNamespaceId(streamId.getNamespaceId())
+        .setCategory(Constants.Notification.Stream.STREAM_FEED_CATEGORY)
+        .setName(String.format("%sSize", streamId.getName()))
+        .build();
+    }
+
+    /**
+     * Cancel the currently scheduled stream size polling task, and reschedule one for later.
+     */
+    private void cancelPollingAndScheduleNext() {
```

Comment:
I think this method must be synchronized. Because it can be invoked concurrently by multiple threads. Both would cancel the existing one, and both would create a new polling runnable and schedule it. Then you will poll twice.